pkg/controllers/leaderworkerset_controller.go

```diff
@@ -200,6 +181,139 @@ func SetupIndexes(indexer client.FieldIndexer) error {
 	})
 }
 
+// leaderTemplate() here indicates the leaderTemplate or the workerTemplate
+// if leaderTemplate is nil.
+// - if leader template updated, we'll roll out the whole lws.
+// - if leader template is nil, and worker template updated, we'll roll out the whole lws.
+// - if leader template is nil, but we update with
+
+// Rolling update will always wait for the former replica ready then process the next one,
+// we didn't consider rollout strategy type here for we only support rollingUpdate now,
+// once we have more policies, we should update the logics here.
+// Possible scenarios:
+//   - When sts is under creation, partition is always 0 because pods are created in parallelism, rolling update is not fit here.
+//   - When sts is rolling update, the partition will start from the last index to the 0 index step by maxUnavailable.
+//   - When sts is rolling update, replicas scaled up, the partition will not change until scaling ended.
+//   - When sts is rolling update, replicas scaled down, the partition will not change unless (replicas-scaleDownedSize) < partition,
+//     partition will be reset to (replicas-scaleDownedSize).
+//
+// When all set, the partition should always be zero.
+func (r *LeaderWorkerSetReconciler) rollPartition(ctx context.Context, lws *leaderworkerset.LeaderWorkerSet) (int32, error) {
+	sts := &appsv1.StatefulSet{}
+	err := r.Get(ctx, types.NamespacedName{Name: lws.Name, Namespace: lws.Namespace}, sts)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return 0, err
+	}
+
+	// If sts not created yet, all partitions should be updated.
+	if apierrors.IsNotFound(err) {
+		return 0, nil
+	}
+
+	partition := *sts.Spec.UpdateStrategy.RollingUpdate.Partition
+	replicas := *lws.Spec.Replicas
+	rollingStep, err := intstr.GetValueFromIntOrPercent(&lws.Spec.RolloutStrategy.RollingUpdateConfiguration.MaxUnavailable, int(replicas), false)
+	if err != nil {
+		return 0, err
+	}
+
+	// Indicates a new rolling update here.
+	if templateUpdated(sts, lws) {
+		if replicasUpdated(sts, lws) {
+			return min(*lws.Spec.Replicas, *sts.Spec.Replicas), nil
+		} else {
+			return replicas - int32(rollingStep), nil
+		}
+	}
+
+	// Otherwise always return 0.
+	if partition == 0 {
+		return 0, nil
+	}
+
+	stsSelector := client.MatchingLabels(map[string]string{
+		leaderworkerset.SetNameLabelKey: lws.Name,
+	})
+	var stsList appsv1.StatefulSetList
+	if err := r.List(ctx, &stsList, stsSelector, client.InNamespace(lws.Namespace)); err != nil {
+		return 0, err
+	}
+
+	var finishedPart int32
+	templateHash := utils.LeaderWorkerTemplateHash(lws)
+stop:
+	for i := replicas - 1; i >= 0; i-- {
+		for j := len(stsList.Items) - 1; j >= 0; j-- {
```

Comment:
qq, why we traverse the stsList from the last to the first one as well?

Reply:
Because at first I though the items will start from the last index to 0, which will accelerate the calculation, but seems not that case, but I didn't change the implementation for lazy reason.
Anyway I think it doesn't make any difference? 😅